File: benchmarks/58-nonlinear-systems/nonlinear_systems.py
#!/usr/bin/env python3
"""Experiment 58 -- Nonlinear systems (scipy.optimize column).

Same kernels as ``nonlinear_systems.m``, same order and labels.

Baseline is compiled (scipy.optimize.fsolve -> MINPACK hybrd/hybrj).  Each
system has a unique root near the shared start, so both solvers land on the same
solution and the check (a coordinate or the sum of the solution) agrees.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import math
import numpy as np
from scipy.optimize import fsolve

from harness import bench, check, require

require(["scipy.optimize:fsolve"])


def r6(x):
    return int(math.floor(float(x) * 1e6 + 0.5))


# ---- small systems ------------------------------------------------------

def poly2(v):
    return [v[0] ** 2 + v[1] ** 2 - 4, v[0] * v[1] - 1]

bench("FindRoot 2x2 polynomial", lambda: fsolve(poly2, [1.5, 0.5]))
check("FindRoot 2x2 polynomial", r6(fsolve(poly2, [1.5, 0.5])[0]))


def transc2(v):
    return [math.cos(v[0]) - v[1], math.sin(v[1]) - v[0] / 2]

bench("FindRoot 2x2 transcendental", lambda: fsolve(transc2, [1.0, 1.0]))
check("FindRoot 2x2 transcendental", r6(fsolve(transc2, [1.0, 1.0]).sum()))


def bf3(v):
    x, y, z = v
    return [3 * x - math.cos(y * z) - 0.5,
            x ** 2 - 81 * (y + 0.1) ** 2 + math.sin(z) + 1.06,
            math.exp(-x * y) + 20 * z + (10 * math.pi - 3) / 3]

bench("FindRoot Burden-Faires 3x3", lambda: fsolve(bf3, [0.1, 0.1, -0.1]))
check("FindRoot Burden-Faires 3x3", r6(fsolve(bf3, [0.1, 0.1, -0.1]).sum()))


# ---- Broyden tridiagonal, N=10 and N=40 ---------------------------------

def broyden_resid(v):
    n = len(v)
    f = np.empty(n)
    for i in range(n):
        left = v[i - 1] if i > 0 else 0.0
        right = v[i + 1] if i < n - 1 else 0.0
        f[i] = (3 - 2 * v[i]) * v[i] - left - 2 * right + 1
    return f


bench("FindRoot Broyden tridiagonal N=10", lambda: fsolve(broyden_resid, -np.ones(10)))
check("FindRoot Broyden tridiagonal N=10", r6(fsolve(broyden_resid, -np.ones(10)).sum()))

bench("FindRoot Broyden tridiagonal N=40", lambda: fsolve(broyden_resid, -np.ones(40)))
check("FindRoot Broyden tridiagonal N=40", r6(fsolve(broyden_resid, -np.ones(40)).sum()))

File: benchmarks/58-nonlinear-systems/nonlinear_systems.m
(* ==========================================================================
   Experiment 58 -- Nonlinear systems (group E)
   ==========================================================================
   WHAT IT MEASURES.  Multivariate FindRoot -- src/numerical_roots/findroot.c
   and nsolve_system.c -- solving systems of nonlinear equations from a starting
   point, the vector analogue of the scalar root-finding in experiment 14.  Two
   small systems (a polynomial and a transcendental), the Broyden tridiagonal
   test at N=10 and N=40 (a scaling row for the Jacobian solve), and the
   Burden-Faires 3x3 system.

   Baseline is scipy.optimize.fsolve (MINPACK hybrd/hybrj, compiled).  Each
   system has a unique root near the shared starting point, so both solvers land
   on the same solution and the check -- a coordinate or the sum of the solution
   -- agrees.

   Note the plain-symbol requirement: FindRoot rejects indexed variables xb[i],
   so the N-variable systems build distinct symbols v1..vN and inject the
   equations/spec with Evaluate (FindRoot holds its arguments).
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"FindRoot"}];

(* ---- small systems ----------------------------------------------------- *)

bench["FindRoot 2x2 polynomial", FindRoot[{x^2 + y^2 == 4, x y == 1}, {{x, 1.5}, {y, 0.5}}];];
check["FindRoot 2x2 polynomial",
      Round[10^6 (x /. FindRoot[{x^2 + y^2 == 4, x y == 1}, {{x, 1.5}, {y, 0.5}}])]];

bench["FindRoot 2x2 transcendental",
      FindRoot[{Cos[x] - y == 0, Sin[y] - x/2 == 0}, {{x, 1.0}, {y, 1.0}}];];
check["FindRoot 2x2 transcendental",
      Round[10^6 ({x, y} /. FindRoot[{Cos[x] - y == 0, Sin[y] - x/2 == 0},
                                     {{x, 1.0}, {y, 1.0}}]) . {1, 1}]];

bench["FindRoot Burden-Faires 3x3",
      FindRoot[{3 x - Cos[y z] - 1/2 == 0,
                x^2 - 81 (y + 1/10)^2 + Sin[z] + 106/100 == 0,
                Exp[-x y] + 20 z + (10 Pi - 3)/3 == 0},
               {{x, 1/10}, {y, 1/10}, {z, -1/10}}];];
check["FindRoot Burden-Faires 3x3",
      Round[10^6 ({x, y, z} /. FindRoot[{3 x - Cos[y z] - 1/2 == 0,
                x^2 - 81 (y + 1/10)^2 + Sin[z] + 106/100 == 0,
                Exp[-x y] + 20 z + (10 Pi - 3)/3 == 0},
               {{x, 1/10}, {y, 1/10}, {z, -1/10}}]) . {1, 1, 1}]];

(* ---- Broyden tridiagonal, N=10 and N=40 -------------------------------- *)
(* Built inline with distinct plain symbols p1..p10 / q1..q40 (FindRoot rejects
   indexed variables, and a Module wrapper around Symbol[] recurses). *)

v10 = Table[Symbol["p" <> ToString[i]], {i, 10}];
e10 = Table[(3 - 2 v10[[i]]) v10[[i]] - If[i > 1, v10[[i - 1]], 0]
            - 2 If[i < 10, v10[[i + 1]], 0] + 1 == 0, {i, 10}];
s10 = Transpose[{v10, ConstantArray[-1., 10]}];
bench["FindRoot Broyden tridiagonal N=10", FindRoot[Evaluate[e10], Evaluate[s10]];];
check["FindRoot Broyden tridiagonal N=10",
      Round[10^6 Total[v10 /. FindRoot[Evaluate[e10], Evaluate[s10]]]]];

v40 = Table[Symbol["q" <> ToString[i]], {i, 40}];
e40 = Table[(3 - 2 v40[[i]]) v40[[i]] - If[i > 1, v40[[i - 1]], 0]
            - 2 If[i < 40, v40[[i + 1]], 0] + 1 == 0, {i, 40}];
s40 = Transpose[{v40, ConstantArray[-1., 40]}];
bench["FindRoot Broyden tridiagonal N=40", FindRoot[Evaluate[e40], Evaluate[s40]];];
check["FindRoot Broyden tridiagonal N=40",
      Round[10^6 Total[v40 /. FindRoot[Evaluate[e40], Evaluate[s40]]]]];
